src/index.rs

```diff
@@ -166,21 +166,22 @@ impl FromStr for Index {
         } else if s.starts_with('0') && s != "0" {
             Err(ParseIndexError::LeadingZeros)
         } else {
-            let idx = s.parse::<usize>().map(Index::Num)?;
-            if s.chars().all(|c| c.is_ascii_digit()) {
-                Ok(idx)
-            } else {
-                // this comes up with the `+` sign which is valid for
-                // representing a `usize` but not allowed in RFC 6901 array
-                // indices
-                let mut invalid: Vec<_> = s.chars().filter(|c| !c.is_ascii_digit()).collect();
-                // remove duplicate characters
-                invalid.sort_unstable();
-                invalid.dedup();
-                Err(ParseIndexError::InvalidCharacters(
-                    invalid.into_iter().collect(),
-                ))
-            }
+            s.chars().position(|c| !c.is_ascii_digit()).map_or_else(
+                || {
+                    s.parse::<usize>()
+                        .map(Index::Num)
+                        .map_err(ParseIndexError::from)
```

Comment:
This should almost never occur now, the only case I think it might is if the number is too large.